Typed Properties

nikic opened the pull request:
A place to leave review comments...

krakjoe committed:
better detection

krakjoe committed:
caches and things ... Thanks Dmitry :)

krakjoe committed:
fix foreach, possibly ...

krakjoe committed:
fix foreach
clean non persistent type info

krakjoe committed:
change decl groups

krakjoe committed:
overflow consistency

krakjoe committed:
moar consistent ... even if strange

krakjoe committed:
make var_dump more acceptable

krakjoe committed:
better

krakjoe committed:
Merge branch 'master' of https://github.com/php/php-src into typed-properties

Conflicts:
	Zend/zend_vm_execute.h

krakjoe committed:
haaaaaands

krakjoe committed:
add overflowingness test

krakjoe committed:
make travis run

krakjoe committed:
fix test [ci skip]

krakjoe committed:
fix memory error in std_write_property

krakjoe committed:
fix another possible memory error

krakjoe committed:
better guard here and use _ex to avoid ever reading type info, we know its an object

krakjoe committed:
don't perform check on inline properties on read, only for magic
fix binary op handler
add case to zend_vm_is_fetching_reference
optimize zend_verify_property_type for most common case in first branch
remove useless undef

krakjoe committed:
use more efficient macro when possible

krakjoe committed:
fix test case given by dmitry ...

krakjoe committed:
only need null check here

krakjoe committed:
use more efficient stuff here

krakjoe committed:
this was checked on write too

krakjoe committed:
safety for non standard objects, remove pointless check

krakjoe committed:
some more improvements ... haaaands

krakjoe committed:
use api stuff here

krakjoe committed:
catch more cases here

krakjoe committed:
nothin to see here ...

krakjoe committed:
improve performance

krakjoe committed:
cold error path

krakjoe committed:
optimize all the things, thanks @bwoebi

krakjoe committed:
more perf stuff

krakjoe committed:
Merge branch 'master' of https://github.com/php/php-src into typed-properties

Conflicts:
	Zend/zend_execute.c
	Zend/zend_vm_execute.h

krakjoe committed:
dtor correct thing

krakjoe committed:
dtor tmp

krakjoe committed:
check correct variable

bwoebi committed:
Optimize & fix value reassign

bwoebi committed:
Easy optimization: constant assignment to objects will always succeed after it succeeded at least once

bwoebi committed:
Optimize property fetching and fix edge cases with __get()

bwoebi committed:
Disallow functions to set references and bad types

krakjoe committed:
missing colons

dstogov committed:
Merge branch 'master' into typed-properties

* master:
  Fixed white-spaces

dstogov committed:
Fixed white-spaces

dstogov committed:
Revert insignificant formatting changes to minimize the patch against master

bwoebi committed:
Fix passed zval reassignments (opcache.protect_memory=1)

bwoebi committed:
Respect also immutable values

dstogov committed:
Added tests

dstogov committed:
Added test (error message should be the same independently of called function)

dstogov committed:
Fixed Zend/tests/type_declarations/typed_properties_044.phpt (error message should be the same independently of called function)

dstogov committed:
Fixed memory leaks

dstogov committed:
Avoid type check for dynamic properties

dstogov committed:
Removed "safety" code for non standard zend objects.
If non-standard objects implement typed properties, they should care about type compatibility checks their seves.

dstogov committed:
Fixed memory leaks

dstogov committed:
Fixed test (NULL is no compatible with "int" and should produce a error)

dstogov committed:
Get rid of zend_verify_scalar_property_type(), use common zend_verify_scalar_property_type() instead.

dstogov committed:
Optimized zend_verify_property_type()

dstogov committed:
Avoid type check for dynamic properties

dstogov committed:
Added support for nullable typed properties

dstogov committed:
Added test

bwoebi committed:
Fix further invalid source zval overwrites

dstogov committed:
Merge branch 'master' into typed-properties

* master:
  Fix memory leak in imagescale()
  Update NEWS
  Better fix for bug #72135
  Fixed bug #72213 (Finally leaks on nested exceptions).
  Forbid dynamic calls to scope introspection functions
  Allow empty property names
  Ensure no entry predecessors for SSA construction
  Replace BB end with BB len
  Fixed bug #72227: imagescale out-of-bounds read
  Fix bug #72241: get_icu_value_internal out-of-bounds read
  Fix bug #72135 - don't create strings with lengths outside int range
  set versions
  Add check for string overflow to all string add operations
  Fix bug #72114 - int/size_t confusion in fread
  Updated NEWS
  Fixed bug #71331 - Uninitialized pointer in phar_make_dirstream()

dstogov committed:
Reverted incomplete fix

dstogov committed:
Fixed test

dstogov committed:
A better to fix for source zval separation on type conversion caused by weak property type check.

dstogov committed:
Changed behavior of uninitialized nullable properties (accoding to conversation with Joe)

dstogov committed:
Revert useless changes

dstogov committed:
Removed useless check (references are never fetched using FETCH_OBJ_RW)

dstogov committed:
Removed duplicate code

dstogov committed:
Added support for nullavle types in AST pretty printer

dstogov committed:
Revert useless change

dstogov committed:
Added test (memory leaks)

dstogov committed:
"string" mat be also a valid "callable"

dstogov committed:
Removed unused function and macro

dstogov committed:
Optimization

dstogov committed:
Removed unused zend_object_fetch_property_type_info() and renamed zend_object_fetch_property_type_info_ex() into zend_object_fetch_property_type_info()

dstogov committed:
Fixed wrong exception when passing element of typed property by reference.

dstogov committed:
Optimisation

dstogov committed:
Added test (failed with valgrind)

dstogov committed:
Added test (failed with valgrind)

dstogov committed:
Fixed memory leak

dstogov committed:
Check type after constant resolution

dstogov committed:
Support for type checks of named constants in property default values.

dstogov committed:
Fixed memory leaks

dstogov committed:
Faster +=/++/--

dstogov committed:
Merge branch 'master' into typed-properties

* master:
  update NEWS
  Expose missing flags from libzip at least >= 0.11.x
  Fix typo in phpdbg help.
  Add missing newlines to phpdbg's help for the run command.
  Skip test which is 64bits only
  update NEWS
  fix test portability
  prepare for 5.6.22RC1

dstogov committed:
Fixed ZTS compatibility

dstogov committed:
Fixed compilation warning

dstogov committed:
fixed white-spaces

dstogov committed:
Optimization

dstogov committed:
Cleanup ext/reflection changes

dstogov committed:
Moved BG(unserialize_refs) into php_unserialize_data_t

dstogov committed:
Keep GCC from incorrect over-optimisation

dstogov committed:
Removed useles code (default value should be already properly set in trait property)

dstogov committed:
Don't mess NULL and UNDEF default value back and force. Set proper value in first place.

dstogov committed:
Remove duplicate checks

dstogov committed:
Fixed incorrect argument types

bwoebi committed:
Remove bad checks for simple property inheritance

krakjoe committed:
Merge branch 'master' of https://github.com/php/php-src into typed-properties

Conflicts:
	Zend/zend_vm_gen.php

krakjoe committed:
changes prompted by review

krakjoe committed:
make consistent with RFC (nullables)

krakjoe committed:
fix inheritance error

bwoebi committed:
Merge branch 'master' into typed_properties

bwoebi committed:
Initial ref types patch
Tests and functions outside ext/standard still need patching

bwoebi committed:
Fix make test

bwoebi committed:
Take properly care of cases where the by-ref value is *consumed* first

bwoebi committed:
Remove all the nice code making it actually worse again ...

bwoebi committed:
Use reftype, not its location...

krakjoe committed:
few changes for bob to check

krakjoe committed:
Merge branch 'typed_ref_properties' of https://github.com/bwoebi/php-src into typed_ref_properties

krakjoe committed:
more zpp changes

krakjoe committed:
finish pcntl, probably

krakjoe committed:
some of sockets

krakjoe committed:
more zpppppp stuff

krakjoe committed:
use macros that actually exist, rather than nonsense

bwoebi committed:
Fix a ton of bugs and tests

bwoebi committed:
Add more guards for typed by-ref assign + fixes

bwoebi committed:
Fix a few more tests

bwoebi committed:
typo; s/zend_get_long/zval_get_long/

bwoebi committed:
Fix iterator_035.phpt and pcntl/003.phpt

bwoebi committed:
Remove lchown($file, -5); test

Some operating systems choose to return EINVAL for users (and groups) out of range (e.g. user ids must not be negative) instead of EPERM.
This thus cannot be deterministically tested for all systems - and we are anyway already testing the branch with errno with the inexistent filename (ENOENT) - thus no problem removing it.

bwoebi committed:
Merge branch 'master' into typed_ref_properties

bwoebi committed:
Add support for by-ref foreach

bwoebi committed:
Separate static property assignment out into dedicated assign ops

bwoebi committed:
Merge remote-tracking branch 'origin' into typed_ref_properties

dstogov committed:
Fixed optimizer support for ZEND_ASSIGN_OBJ_REF

dstogov committed:
Implemented type inference support for typed properties.

bwoebi committed:
Merge branch 'master' into typed_ref_properties

bwoebi committed:
Merge remote-tracking branch 'github/typed_ref_properties' into typed_ref_properties

dstogov committed:
Fixed tests (made them 32/64-bit independent)

bwoebi committed:
Fix a bunch of bugs related to e8109e09

dstogov committed:
Merge branch 'master' into typed_ref_properties

* master: (42 commits)
  Increase wait time for tests involving sockets to improve Travis relability
  Remove redundant reference handling
  Fixed bug #73916 (zend_print_flat_zval_r doesn't consider reference)
  update NEWS
  Fixed #72898 PHP_FCGI_CHILDREN is not included in phpinfo()
  Fix memory error in reflection export
  Fix bug #69899
  update NEWS
  Fixed bug #73374
  Fixed bug #73904 php-cgi fails to load -c specified php.ini file
  Fixed bug #73904 php-cgi fails to load -c specified php.ini file
  Fixed bug #72680 gdbinit print_cvs not prepared for PHP 7
  Update to SQLite 3.16.2
  added a note to opache.fast_shutdown why its not enabled by default in php.ini
  Fixed bug #71219 configure script incorrectly checks for ttyname_r
  Fixed bug #67707 IV not needed for ECB encryption mode, but it returns a warning
  Add support for "xfail" to phpt SKIPIF sections
  Fixed #73907 (nextSibling property not included in var_dump of DOMNode)
  Fixed bug #61858 (DOMAttr debug info generates E_WARNING)
  Fix glob-wrapper.phpt to not fail in Windows
  ...

bwoebi committed:
Merge branch 'master' of https://github.com/php/php-src into typed_ref_properties

bwoebi committed:
Fix crash in pcntl_wait()

dstogov committed:
Merge branch 'master' into typed_ref_properties

* master:
  Fix bug #73634 - grapheme_strpos illegal memory access
  Fix typo
  Fix bug #73939 memory allocation in signal handler when HAVE_STRUCT_SIGINFO_T
  Add test ini_set with invalid time zone
  Fix test
  Update more functions with path check
  move some var declarations to the top for better grouping
  fix AppVeyor failures with Opcache enabled
  Fix small mistake in generator creation introduced in 9a159f37
  Revert "Merge branch 'bugfix_timeout_61471' of https://github.com/axot/php-src into PHP-7.0"
  Fix segfault in `pcntl_signal_handler` (bug #73939)

bwoebi committed:
Fix constant value overwrite

bwoebi committed:
Add support for iterable typed properties

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Fix build failures

nikic committed:
Initial ref type in MAKE_REF_EX

bwoebi committed:
Solve the clashing FETCH_REF|CACHE_SLOT

bwoebi committed:
Fix few tests

bwoebi committed:
Fix cache_slot of ASSIGN_OBJ_REF

bwoebi committed:
Opcache fixes

bwoebi committed:
Fix build failure in exif

bwoebi committed:
Add one more test for iterables by ref

bwoebi committed:
Fix foreach object iteration with holes

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Restore old code for post_incdec as well

nikic committed:
Adjust test with new behavior

The reference is no longer modified, so no problems. Any necessary
type check would occur in the write_property part.

nikic committed:
Fix compiler warning

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Improve default value handling

 * Don't allow null for callable, consistent with everything else.
   Previously null was allowed as default value, but not at runtime...
 * Allow arrays as default for iterable.
 * Print friendlier error message when using null default value,
   suggesting to use ?Type.

nikic committed:
Allow int default value for float properties

nikic commented on Zend/tests/bug69732.phpt:
```diff
@@ -17,15 +17,14 @@ function ret_assoc() {
+  string(3) "bar"
+  [0]=>
+  string(3) "XXX"
+}
```
Why did this test change? The new output looks wrong to me.

nikic committed:
Use not_in_object_context helper

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Fix ASSIGN_OBJ_REF implementation

Line up logic with ASSIGN_REF, fixing various discrepancies.

nikic commented:
```php
<?php
  
class Test {
    public string $x = "x";
}

$test = new Test;
var_dump($test);
$y = "y";
$test->x =& $y;
var_dump($y, $test);

$z = 42;
$y = $z;
var_dump($y, $z, $test);
```

This currently leaks the string that is created when casting 42 to `"42"` for the weak assignment.

nikic committed:
Don't automatically initialize nullable prop to null

Handle them consistently with other types and initialize to
"uninitialized", unless a default value is explicitly given.

nikic committed:
Forbid callable type hint in typed properties

bwoebi committed:
Merge branch 'extra_static_prop_assign_ops' into typed_ref_properties

bwoebi committed:
Merge branch 'typed_ref_properties' of https://github.com/bwoebi/php-src into typed_ref_properties

bwoebi commented:
```
bob$ ./sapi/cli/php -r '$a = new class { public ?string $b = null; }; $a->b[] = 1; var_dump($a);'
object(class@anonymous)#1 (1) {
  ["b"]=>
  array(1) {
    [0]=>
    int(1)
  }
}
```

Promotion of values by indirect modification (array assign) isn't guarded against yet.

bwoebi committed:
Add support for typed static properties (no opcache yet)

bwoebi committed:
Fix assign to typed ref with weak type conversion

Thanks to @nikic for the test

bwoebi committed:
Fix opcache for typed static properties

bwoebi committed:
Fix remaining tests

nikic commented:
```php
<?php
  
class Test {
    public int $val;

    public function __get($name) {
        return "42";
    }
}

$test = new Test;
var_dump($test);
var_dump($test->val);
```

This currently prints `string(2) "42"`, but it should print `int(42)`.

bwoebi commented:
@nikic  That's expected behavior when property values are not set.

```php
<?php

class Test {
    public $val;

    public function __get($name) {
        return "42";
    }
}

$test = new Test;
unset($test->val); // initialized with null, but our typed properties are uninitialized by default
var_dump($test->val);
```

bwoebi commented:
I'm fixing this, needs also a fix for &__get().

bwoebi committed:
Fixing conversion with __get() on uninit typed properties

Thanks @nikic for the first test

nikic commented:
@bwoebi Thanks!

Another behavior question, what should this print?

```php
class Test {
    public int $val;
}

$test = new Test;
var_dump($test->val = "42"); // string(2) "42"
var_dump($test->val); // int(42)
```

Currently the return value of the assignment is the original string, even though the actually assigned value is an integer.

Is this the behavior we want? We have at least some precedent for returning the "actually assigned" value here:

```php
$str = "foo";
var_dump($str[0] = 1); // string(1) "1"
```

bwoebi committed:
Prevent constant propagation into ASSIGN_*_REF OP_DATA

bwoebi commented:
I'm a little unsure, because, given __set(), we then don't fetch the __get() value back, but just assign it. ... but let's consider that magic which is magic :-P

It actually already sometimes works this way:
```
class Test {
    public int $val = 0;
}

$test = new Test;
$int = &$test->val;
var_dump($int = (string) 42); // int(42)
var_dump($int); // int(42)
```

because for TMPs and VARs, it's modified inline and the original value isn't preserved at all.

So … to avoid unnecessary copying I think it's best to just align it with the assigned value.

However it can be a bit counterintuitive:

```
// I expect $foo and $bar to be both assigned int(42) and eventual weak casts being applied afterwards
$foo = $bar = 42;
```

In particular, if $bar is of type float and $foo of type int. I'm pretty sure that will be quite a wtf moment. (given strict types)
Should be possibly rare enough to not actually matter, but … don't know.

We just definitely shouldn't have mixed behavior, but I can't really tell what's better.

nikic commented:
@bwoebi I think there are some confusing cases either way. My vote goes to returning the converted value.

Furthermore, in the case of reference assignment, I don't think we have a choice in the matter:

```php
class Test {
    public int $x;
}

$test = new Test;
$x = "42";
var_dump($test->x =& $x);
var_dump($test->x);
```

Here `$test->x =& $x` has to be `int(42)`, otherwise the return value couldn't be a reference.

bwoebi commented:
Indeed, I'd bet `$a = &$b` would actually return the deref'ed value, but apparently not.

So, I then agree with your choice. Will work on it then.

bwoebi committed:
Add ZEND_PRE/POST_INC/DEC_STATIC_PROP

bwoebi commented on Zend/zend_execute.c:
```diff
@@ -1607,6 +1607,135 @@ static zend_never_inline void zend_assign_to_string_offset(zval *str, zval *dim,
 	}
 }
 
+static void zend_pre_incdec_property_zval(zval *prop, zend_property_info *prop_info, int inc OPLINE_DC EXECUTE_DATA_DC)
```
Not sure whether it's worth inlining it - I don't think so...

nikic commented:
@bwoebi So one case that I find really concerning regarding the behavior is this one:

```php
class Test {
    public int $x = 42;
}
$test = new Test;
$array = [&$test->x];
unset($test);
// var_dump($array[0]); // Toggle me (unref)
$array[0] = "not an int";
```

I think that this will be a big WTF moment. We can get rid of that particular unref (and I think I will submit a PR for that and see what dmitry says), but this doesn't really solve the fundamental issue.

So I'm wondering if instead we could go with the spirit of "rc=1 reference is not a reference" and not enforce type checks on rc=1 references. Or more specifically to null out the type if an assignment to an rc=1 reference is made.

I know that this is also not fully consistent (because an rc=1 ref can go back to rc=2), but I think it will make more sense than what we do now ... maybe? What do you think?

bwoebi commented:
Yeah :-D

Similarly when we have multiple references:

```
class A {
    public int $x;
}
class B {
    public $x;
}
$a = new A;
$b = new B;
$a->x = &$b->x;
$c = &$b->x;
unset($a);
// now we cannot assign an integer to $b->x or $c
```

This is quite an inherent limitation unless we now start tracking the typed reference holders and their associated type. We definitely _could_ do that and have an extra overhead on reference assign to typed props and reference removal.

nikic commented:
This code currently leaks:

```php
class A {}
class B extends A {}

class Test {
    public A $A;
    public B $B;
}

$test = new Test;
$test->B = new B;
$test->A =& $test->B;
try {
    $test->A = new A;
} catch (TypeError $e) {
    echo "{$e->getMessage()}\n";
}
```

nikic committed:
Property compute intersection types for $obj->prop =& $ref

Test currently fails due to unrelated leaks.

nikic commented:
> This is quite an inherent limitation unless we now start tracking the typed reference holders and their associated type. We definitely could do that and have an extra overhead on reference assign to typed props and reference removal.

I think that the case with multiple references is not really a problem. It's just how we define the semantics: We say that the reference itself has an intrinsic type that is not coupled to the lifetime of any typed property.

The thing that's problematic about the rc=1 case is that the type might go away at any moment due to some innocuous operation. We generally try to treat rc=1 reference like they aren't references, so I think we may want to do this with regards to types as well.

nikic committed:
Add missing test file...

bwoebi commented:
Ah. Can't we just unref then when we encounter rc=1 refs as we anyway have to check for rc=1?

nikic committed:
Fix leak

bwoebi committed:
$a->b = $c must return (string) $c if $a->b is of type string

bwoebi commented:
This last commit (a450c28534fc465bcc833ad3ecfc5fa6a735cf97) changes the signature of write_property, as we want to get the set value and not actually the value a getter will read (there might be writeonly in future or extensions doing things, don't know).

nikic committed:
Drop overeager assert

Function may be used in cases where types are not set.

nikic commented:
@bwoebi Reference type not enforced here:

```php
<?php
  
class Test {
    public static $x;
    public static int $y = 0;
}

Test::$x =& Test::$y;

$rc = new ReflectionClass('Test');
$rc->setStaticPropertyValue('x', 'foo');
var_dump(Test::$y);
```

setStaticPropertyValue does a DEREF+assign right now ^^

bwoebi committed:
Enforce reference checks on ReflectionClass::setStaticPropertyValue() + test

bwoebi committed:
Use coercive mode for setStaticPropertyValue()

This is consistent with behavior of ReflectionProperty::setValue() and all other internal function calling

bwoebi committed:
Keep track of reference types

bwoebi committed:
Fix memory leaks & stupid bugs introduced in last commit

bwoebi committed:
Fix static property ref assignment without types

nikic commented:
Reference type not enforced:

```php
<?php
  
class Test {
    public int $x = 0;
}

$test = new Test;
$x =& $test->x;
extract(['x' => 'foo']);
var_dump($test->x);
```

bwoebi committed:
extract() must check typed refs when overwriting

Thanks @nikic for testing this

nikic commented:
```php
<?php
  
$x = 42;

class Test {
    public int $x;
    public function &__get($name) {
        return $GLOBALS['x'];
    }
    public function __destruct() {
        $GLOBALS['x'] = 'foobar';
    }
}

var_dump((new Test)->x); // string(6) "foobar"
```

nikic commented:
Reference type not enforced:

```php
<?php
  
class Test {
    public int $x = 42;
}

$test = new Test;
preg_match('/foo/', 'foo', $test->x);
var_dump($test->x);
```

nikic committed:
Mark subpats as ARRAY_ASSIGNABLE

nikic committed:
Use TRY_ASSIGN_LONG for zcount parameters

nikic commented:
PCRE issues fixed.

```php
<?php

class Test {
    public float $x = 42.0;
}

$test = new Test;
settype($test->x, 'int');
var_dump($test->x); // currently int
```

nikic commented:
I'm getting quite a few results with `git grep "zend_parse_parameters.*z/"`, some of those at least are legit issues.

bwoebi committed:
Fix refcounting of zend_try_assign()

nikic commented:
I've opened https://github.com/php/php-src/pull/3313 upstream. Leaving this open for our own comments.

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Enforce reference type in settype()

nikic commented:
@bwoebi Reference type not enforced here:

```php
class Test {
    public static int $x = 0;
}

class Test2 extends Test {
}

$x =& Test::$x;
$x = "foo";
var_dump(Test::$x);
```

Static properties use references as part of the implementation :/

nikic commented:
@bwoebi Issue with tracking sources while cloning typed references:

```php
class Test {
    public int $x = 42;
}

$test = new Test;
$x =& $test->x;
$test2 = clone $test;
unset($test);
$x = "foo";
var_dump($test2->x);
```

nikic committed:
Make ZEND_TRY_ASSIGN_* implementation less magic

Should also fix the windows build.

nikic committed:
Unmangle property name before printing

We could also use the name from op1 (what we do for non-static
properties), but as it is conditionally fetched, just unmangling
the name is simpler.

bwoebi committed:
Fix reference type tracking on clone

bwoebi committed:
Merge branch 'master' of https://github.com/php/php-src into typed_ref_properties

nikic committed:
Update extensions for ZEND_TRY_ASSIGN_*

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic commented:
Some TODOs:

 * [x] PDO / mysqli by-reference bound params need to be reviewed.
 * [x] There's a `z/` in xmlrpc that I ignored.
 * [ ] Opcache needs some more changes. (@nikic)
 * [x] Switching to https://wiki.php.net/rfc/typed_properties_v2#references_without_intrinsic_type
 * [x] Handling of array promotion.
 * [x] Add ReflectionProperty::isInitialized()

nikic committed:
Opcache handling fixes and improvements

nikic committed:
Merge remote-tracking branch 'php-src/master' into typed_ref_properties

nikic commented:
This currently throws "Internal error: Could not find the property Test::x".

```php
class Test {
    public static int $x;
}

$rp = new ReflectionProperty('Test', 'x');
var_dump($rp->getValue());
```

nikic committed:
Reduce ASSIGN_STATIC_PROP_REF specialization

This operation should be very rare... At least drop the OP_DATA
specialization.

nikic committed:
Accept BP type in get_static_property

And move the initialization check in there, so it's automatically
handled everywhere using this function.

nikic commented:
Build failure is one of the opcache problems I didn't get around to yet -- exposed by reducing specialization.

bwoebi committed:
Prevent type violations via array auto-promotion

bwoebi committed:
Fix type check for array promotion check

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic commented:
```php
class Test {
    public static int $x;
}

$rp = new ReflectionProperty('Test', 'x');
$rp->setValue("foo");
var_dump($rp->getValue());
```

nikic committed:
Restore fast-path check for uninitialized static property

It's not sufficient to check on the slow-path only, as opcache
may combine cache entries.

bwoebi committed:
Add ReflectionProperty::isInitialized()

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Minor cleanups

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -5669,6 +5669,62 @@ ZEND_METHOD(reflection_property, setValue)
+	}
+
+	if (ref->prop.flags & ZEND_ACC_IMPLICIT_PUBLIC) {
+		RETURN_TRUE;
```
This check doesn't look right to me for a case like the following:

```php
$obj1 = (object)['foo' => 'bar'];
$rp = new ReflectionProperty($obj1, 'foo');
$obj2 = (object)[];
var_dump($rp->isInitialized($obj2)); // bool(true) instead of bool(false)
```

bwoebi committed:
Fix refcount on std_write_property with coercion

nikic committed:
Fix ReflectionProperty::isInitialized() for dynamic props

A dynamic properties may exist at time of ReflectionProperty
instantiation but be unset later, so we cannot short-circuit.

nikic committed:
Fix ReflectionProperty::setValue() on typed static property

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Fix class check

bwoebi committed:
References are the intersection of all present property types at any time

bwoebi committed:
Fix a last test

nikic committed:
Remove 't' zpp type

The 't' type will not be usable by extensions.

Replace it with zend_try_array_init(_size) functions.

nikic committed:
Drop ARRAY_ASSIGNABLE as well

nikic committed:
Don't constrain type intersections

This allows intersecting ?int and ?string with value null, and is
conceptually forward-compatible with union types, at which point we
must deal with the problem of conflicting coercive types anyway.
Implemented in a way that only performs type coercion (if necessary)
once.

nikic committed:
Merge branch 'master' into typed_ref_properties

nikic committed:
Abort early on exception in extract()

In line with upstream. Also remove silent parameter from
zend_try_assign_ex().

nikic committed:
Move throwing of error into verify_ref_assignable_zval()

bwoebi committed:
Merge branch 'master' of https://github.com/php/php-src into typed_ref_properties

bwoebi committed:
Fix build without execute_data $register

bwoebi committed:
Move push/pop function prologue off fast path in VM

nikic commented:
This code should generate an error due to incompatible properties (also needs to be mentioned in the RFC):

```php
trait T1 {
    public int $prop;
}
trait T2 {
    public string $prop;
}
class C {
    use T1, T2;
}
```

nikic committed:
Resolve aliases during property type LSP checks

nikic committed:
Check for conflicting property types in traits

bwoebi committed:
Fix write_property handlers and add zend_std_get_property_with_info

bwoebi committed:
Remove *tmp arg from zend_verify_property_type()

bwoebi committed:
Remove useless separations/copies

bwoebi committed:
handle by ref mysqli/PDO statement result binding

bwoebi committed:
Fix string length updates

bwoebi committed:
Merge branch 'master' into typed_ref_properties

bwoebi committed:
Fix a last function to use zend_try_assign()

bwoebi committed:
Fix non-string key usage with typed properties

bwoebi committed:
Merge remote-tracking branch 'origin/master' into typed_ref_properties

bwoebi committed:
Merge remote-tracking branch 'origin/master' into typed_ref_properties

petk committed:
Add directory structure to the contributing docs

This patch adds a php-src directory structure introduction to the
contributing guidelines. Thanks to the discussion at [1] and PHP
manual.

[1] https://github.com/php/php-src/pull/3520

weltling committed:
Upgrade PCRE2 to 10.32

(cherry picked from commit d918e0776b5168aed2707b0ca500589844f0faa8)

petk committed:
Remove HAVE_TIME_H

The `<time.h>` header file is part of the standard C89 headers [1] and
on current systems can be included unconditionally.

Since PHP requires at least C89 or greater, the `HAVE_TIME_H` symbol
defined by Autoconf in ext/pdo_sqlite/config.m4 [2] can be ommitted and
simplifed.

Additionally, since PHP didn't define `HAVE_TIME_H` prior in the
configure.ac the occurrence of this symbol in cli can be removed.

Refs:
[1] https://port70.net/~nsz/c/c89/c89-draft.html#4.1.2
[2] https://git.savannah.gnu.org/cgit/autoconf.git/tree/lib/autoconf/headers.m4

petk committed:
Fix permissions

Certain files don't need to be executables.

dstogov committed:
Improved hash copying

dstogov committed:
Use macro

cmb69 committed:
Update to SQLite 3.25.1

dstogov committed:
Fixed bug #76711 (OPcache enabled triggers false-positive "Illegal string offset")

dstogov committed:
Merge branch 'PHP-7.3'

* PHP-7.3:
  Fixed bug #76711 (OPcache enabled triggers false-positive "Illegal string offset")

sebastianbergmann committed:
Fix

sebastianbergmann committed:
Fix

cmb69 committed:
Update NEWS and UPGRADING wrt. PR 2760

cmb69 committed:
Merge branch 'PHP-7.3'

* PHP-7.3:
  Fix
  Fix

weltling committed:
Fixed bug #76909 preg_match difference between 7.3 and < 7.3

Applied upstream patch, see https://bugs.exim.org/show_bug.cgi?id=2321

weltling committed:
Add test for bug #76909

weltling committed:
Merge branch 'PHP-7.3'

* PHP-7.3:
  Add test for bug #76909
  Fixed bug #76909 preg_match difference between 7.3 and < 7.3
  Upgrade PCRE2 to 10.32

weltling committed:
[ci skip] Update UPGRADING.INTERNALS

petk committed:
Rename *.p5c files to *.inc

When including files in PHP tests a good practice so far has been to use
the *.inc extension. This patch renames few *.p5c files that are
included in phpt files.

weltling committed:
Fix test and add skip section

weltling committed:
Merge branch 'PHP-7.3'

* PHP-7.3:
  Fix test and add skip section

nikic committed:
Merge remote-tracking branch 'php-src/master' into typed_ref_properties

bwoebi committed:
Add test for typed properties with integer keys

nikic commented:
Closing this one in favor of the php-src PR.
